src/api/cpp/nixl_v2_1.h

```diff
@@ -0,0 +1,232 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * @file nixl.h (NVIDIA Inference X Library)
+ * @brief These are NIXL Core APIs for applications
+ */
+#ifndef _NIXL_H
+#define _NIXL_H
+
+#include <vector>
+#include <string>
+#include <cstdint>
+#include <memory>
+#include <functional>
+
+/* notes:
+    * Connections are handled implicitly by nixl backend
+    * Memory registrations are handled implicitly by nixl backend
+*/
+
+namespace nixl {
+
+    /* 
+        a nixl pod is a group of nixl agents that can transfer memory between each other.
+        the vision is of centralized meta data controller.
+        nixl agents communicate meta data only via the centralized controller.
+        the meta data messages are defined in the pod::msg namespace.
+        the Client class is intantiated in the Agent class - this is the Agent interface with the centralized controller.
+    */
+    namespace pod {
+        namespace obj {
+            enum class Type {
+                AGENT,
+                MEMSPACE
+            };
+
+            struct Agent {
+                std::string ip;
+                uint16_t    port;
+            };
+
+            struct MemSpace {
+                struct Agent agent;
+                uint32_t     key;
+            };
+        } // namespace obj
+
+        namespace msg {
+            enum class Opcode {
+                ADD_AGENT,
+                DEL_MEMSPACE
+            };
+
+            struct AddAgent {
+                struct obj::Agent agent;
+            };
+
+            struct DelAgent {
+                struct obj::Agent agent;
+            };
+
+            struct AddMemSpace {
+                struct obj::MemSpace           memSpace;
+                uint64_t                       len;
+                std::vector<struct obj::Agent> accessList;
+            };
+
+            struct DelMemSpace {
+                struct obj::MemSpace memSpace;
+            };
+
+            struct Msg {
+                Opcode                 opcode;
+                union {
```

Comment:
Unions in C++ with non-POD typed members have terrible ergonomics. Just try to make this simple example work: https://www.programiz.com/online-compiler/5TjIfXfqAW48k

Reply:
Agree, std::variant would be much more ergonomic to  use.